Lib/pathlib.py

```diff
@@ -1385,6 +1386,116 @@ def expanduser(self):
+        to abort the walk.  Note that the filename is available as the
+        filename attribute of the exception object.
+
+        By default, Path.walk does not follow symbolic links to subdirectories
+        on systems that support them.  In order to get this functionality, set
+        the optional argument 'follow_symlinks' to true.
+
+        Caution: if self is a relative Path, don't change the
+        current working directory between resumptions of walk. walk never
+        changes the current directory, and assumes that the client doesn't
+        either.
+        
+        Caution: walk assumes the directories have not been modified between
+        its resumptions. I.e. If a directory from dirnames has been replaced
+        with a symlink and follow_symlinks=False, walk will still try to
+        descend into it. To prevent such behavior, remove directories from
+        dirnames if they have been modified and you do not want walk to 
+        descend into them anymore.
+
+        Example:
+
+        from pathlib import Path
+        for root, dirs, files in Path('Lib/concurrent').walk(on_error=print):
+            print(
+                root,
+                "consumes",
+                sum((root / file).stat().st_size for file in files),
+                "bytes in",
+                len(files),
+                "non-directory files"
+            )
+            # don't visit __pycache__ directories
+            if '__pycache__' in dirs:
+                dirs.remove('__pycache__')
+        """
+        sys.audit("pathlib.Path.walk", self, on_error, follow_symlinks)
+        return self._walk(True, on_error, follow_symlinks)
+
+    def walk_bottom_up(self, on_error=None, follow_symlinks=False):
+        """Generate a bottom up directory tree from this directory
+
+        The return type and arguments are identical to Path.walk.
+        
+        Unline walk, the caller cannot modify the dirnames to prune the
+        search or to impose a specific order of visiting because the
+        list of directories to visit is calculated before yielding it.
+        """
+        sys.audit("pathlib.Path.walk_bottom_up", self, on_error, follow_symlinks)
+        return self._walk(False, on_error, follow_symlinks)
+
+    def _walk(self, top_down, on_error, follow_symlinks):
+        dirs = []
+        nondirs = []
+        dir_entry_cache = {}
+
+        # We may not have read permission for self, in which case we can't
+        # get a list of the files the directory contains. os.walk
+        # always suppressed the exception then, rather than blow up for a
+        # minor reason when (say) a thousand readable directories are still
+        # left to visit. That logic is copied here.
+        try:
+            scandir_it = self._scandir()
+        except OSError as error:
+            if on_error is not None:
+                on_error(error)
+            return
+
+        with scandir_it:
+            for entry in scandir_it:
```

Comment:
I replaced while with for. To be honest, I am not sure why we had while in there to begin with. It seems like it was just a piece of legacy that nobody removed when [PEP-471](https://peps.python.org/pep-0471/#notes-on-exception-handling) has been created.

Reply:
I figured it was because calling `next(scandir_it)` can raise? Or is that not the case?

Reply:
That's what I thought too. But then I looked at PEP-471. Look at the link I supplied, read that section, and then ctrl + f "while True". What you'll see is that they use "for" everywhere, even in 'fine grained error handling' example. And they only use while in a **rejected** example. So... It feels to me like scandir only raises errors on its initial call and on method calls, not on next.